OSDOCS#6056: cert-manager: Authenticating the cert-manager Operator with AWS [non-sts] (#59907)
openshift/openshift-docs: modules/cert-manager-configure-cloud-credentials-aws-non-sts.adoc

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-aws.adoc

:_content-type: PROCEDURE
[id="cert-manager-configure-cloud-credentials-aws-non-sts"]
= Configuring cloud credentials for the {cert-manager-operator} on AWS

To configure the cloud credentials for the {cert-manager-operator} on the AWS cluster you must create a `CredentialsRequest` object and allow the Cloud Credential Operator to generate the cloud credentials secret.

.Prerequisites

> * The {cert-manager-operator} 1.11.1 or later is installed.

Reviewer: Similar to https://github.com/openshift/openshift-docs/pull/60433 add one more prerequisite about CCO mode.

. Create a `CredentialsRequest` resource YAML file, for example, `sample-credential-request.yaml`, as follows:
+
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries:
    - action:
      - "route53:GetChange"
      effect: Allow
      resource: "arn:aws:route53:::change/*"
    - action:
      - "route53:ChangeResourceRecordSets"
      - "route53:ListResourceRecordSets"
      effect: Allow
      resource: "arn:aws:route53:::hostedzone/*"
    - action:
      - "route53:ListHostedZonesByName"
      effect: Allow
      resource: "*"
  secretRef:
    name: aws-creds
    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----
+
[NOTE]
====
The use of the `dns.admin` role in the `sample-credential-request.yaml` file is for convenience. To ensure cert-manager runs
under the least privilege service account, you must create a custom role with the following permissions:

* `dns.resourceRecordSets.*`
* `dns.changes.*`
* `dns.managedZones.list`
====

> 

Reviewer: These permissions are for GCP cloudDNS, not for AWS route53.

Reviewer: Ah, found it, Swarup's https://github.com/openshift/openshift-docs/pull/59907#issuecomment-1570414873 is good reminder, but might have been commented in wrong PR, sorry I did not point this out with available time : ).
It is intended to be better added in another GCP PR https://github.com/openshift/openshift-docs/pull/60433 which is already merged.

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

. Update the subscription object for {cert-manager-operator} by running the following command:
+
[source,terminal]
----

> $ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"aws-creds"}]}}}'

Reviewer: ```suggestion
$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type=merge -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"aws-creds"}]}}}'
```

----

.Verification

> . Get the name of a cert-manager controller pod by running the following command:

Reviewer: ```suggestion
. Get the name of the redeployed cert-manager controller pod by running the following command:
```

+
[source,terminal]
----
$ oc get pods -l app.kubernetes.io/name=cert-manager -n cert-manager
----
+
.Example output
[source,terminal]
----
NAME                          READY   STATUS    RESTARTS   AGE
cert-manager-bd7fbb9fc-wvbbt  1/1     Running   0          15m39s
----

> . Verify that AWS credentials are updated by running the following command:

Reviewer: I am referring to non-sts GCP https://github.com/openshift/openshift-docs/pull/60433/files for similar line, found it is:
```
. Verify that the cert-manager controller pod is updated with GCP workload identity credential ...
```
This is wrong. "GCP workload identity" is only for GCP STS, not for non-sts GCP. Sorry for didn't catch this in last reviews. But https://github.com/openshift/openshift-docs/pull/60433 is already merged, could you update that?
For example, by removing "workload identity", update to:
```
. Verify that the cert-manager controller pod is updated with GCP credential ...
```

+
[source,terminal]
----

> $ oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o=jsonpath='{.spec.volumes}'

Reviewer: Like other 3 PRs where there was similar issue, `-o=jsonpath='{.spec.volumes}'` prints json, but the following given output is yaml.

----
+
.Example output
[source,terminal]
----

> - name: cloud-credentials
>     secret:
>       secretName: aws-creds

Reviewer: One more indention.
Also show the corresponding mountPath:
```suggestion
spec:
  containers:
  - args:
    ...
    - mountPath: /.aws
      name: cloud-credentials
  ... 
  volumes:
  ...
  - name: cloud-credentials
    secret:
      ...
      secretName: aws-creds
```

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-aws.adoc

:_content-type: PROCEDURE

> [id="cert-manager-configure-cloud-credentials-aws-non-sts"]

Reviewer 2: ```suggestion
[id="cert-manager-configure-cloud-credentials-aws-non-sts_{context}"]
```

= Configuring cloud credentials for the {cert-manager-operator} on AWS

> To configure the cloud credentials for the {cert-manager-operator} on the AWS cluster you must create a `CredentialsRequest` object and allow the Cloud Credential Operator to generate the cloud credentials secret.

Reviewer 2: If this is what the sentence means
```suggestion
To configure the cloud credentials for the {cert-manager-operator} on the AWS cluster you must generate the cloud credentials secret by creating a `CredentialsRequest` object, and allowing the Cloud Credential Operator.
```

.Prerequisites

> * The {cert-manager-operator} 1.11.1 or later is installed.

Reviewer 2: ```suggestion
* You have installed the {cert-manager-operator} 1.11.1 or later.
```

    namespace: cert-manager
  serviceAccountNames:
  - cert-manager
----

. Create a `CredentialsRequest` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sample-credential-request.yaml
----

. Update the subscription object for {cert-manager-operator} by running the following command:
+
[source,terminal]
----
$ oc -n cert-manager-operator patch subscription openshift-cert-manager-operator --type='merge' -p '{"spec":{"config":{"env":[{"name":"CLOUD_CREDENTIALS_SECRET_NAME","value":"aws-creds"}]}}}'
----

.Verification

. Get the name of a cert-manager controller pod by running the following command:
+
[source,terminal]
----
$ oc get pods -l app.kubernetes.io/name=cert-manager -n cert-manager
----
+
.Example output
[source,terminal]
----
NAME                          READY   STATUS    RESTARTS   AGE
cert-manager-bd7fbb9fc-wvbbt  1/1     Running   0          15m39s
----

. Verify that AWS credentials are updated by running the following command:
+
[source,terminal]
----
$ oc get -n cert-manager pod/<cert-manager_controller_pod_name> -o=jsonpath='{.spec.volumes}'
----
+
.Example output
[source,terminal]
----
- name: cloud-credentials
    secret:

>       secretName: AWS-creds

Reviewer 3: ```suggestion
      secretName: aws-creds
```
AWS is small case here.

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-aws.adoc

:_content-type: PROCEDURE
[id="cert-manager-configure-cloud-credentials-aws-non-sts"]
= Configuring the {cert-manager-operator} to use prepared cloud credentials

You can configure the {cert-manager-operator} to use prepared cloud credentials on AWS Security Token Service cluster by using the `ccoctl` binary.

.Prerequisites

* You must extract and prepare the `ccoctl` binary.

.Procedure

. The {cert-manager-operator} creates the `CredentialsRequest` object in the `openshift-cloud-credential-operator` namespace. Get the role by running the following command:
+
[source,terminal]
----
$ oc get credentialsrequest -n openshift-cloud-credential-operator -o yaml
----
+
.Example output
[source,terminal]
----
arn:aws:iam::<AWS-account-id>:role/abcd-cert-manager
----

. Annotate the service account by running the following command:
+
[source,terminal]
----

> $ oc annotate sa/cert-manager -n cert-manager eks.amazonaws.com/role-arn="arn:aws:iam::<AWS-account-id>:role/abcd-cert-manager"

Reviewer 3: You can skip this annotate step in non-sts AWS, as for most non-STS clusters they won't have the pod identity webhook functional (which is responsible for injecting the credentials into the pod base upon annotations in the service account).
xref: https://redhat-internal.slack.com/archives/C04TMSTHUHK/p1683663618177239?thread_ts=1683658420.220019&cid=C04TMSTHUHK

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-authenticate-aws.adoc

:_content-type: PROCEDURE
[id="cert-manager-configure-cloud-credentials-aws-non-sts"]
= Configuring cloud credentials for the {cert-manager-operator} on AWS

To configure the cloud credentials for the {cert-manager-operator} on the AWS cluster you must create a `CredentialsRequest` object and allow the Cloud Credential Operator to generate the cloud credentials secret.

.Prerequisites

* The {cert-manager-operator} 1.11.1 or later is installed.

.Procedure

. Create a `CredentialsRequest` resource YAML file, for example, `sample-credential-request.yaml`, as follows:
+
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: cert-manager
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries:
    - action:
      - "route53:GetChange"
      effect: Allow
      resource: "arn:aws:route53:::change/*"
    - action:

>       - "route53:ChangeResourceRecordSets",

Reviewer 3: The last comma is not needed, probably a typo.
